text/projects/Lukasz_Selwa/Galil_Seiferas.tex

```diff
@@ -0,0 +1,308 @@
+Niech $v_1, v_2$ to $k$-HRP słowa $w$, $|v_1| < |v_2|$ oraz $[L_1, R_1], [L_2, R_2]$ to zakresy odpowiednio $v_1, v_2$. Wtedy $R_1 < L_2$.
+\end{lemma}
+\begin{proof}
+Niech $z$ będzie prefiksem $w$ długości $R_1$. Jeśli $L_2 \leq R_1$ to z tego wynika, że $v_2^2$ jest prefiksem $z$. Ponieważ, $|v_1| < |v_2|$ i $|v_1|$ jest okresem $z = v_2^2 z'$ to $|v_1|$ jest też okresem $v_2^2$. Z lematu o okresowości wynika, że $\gcd(|v_1|, |v_2|)$ też musi być okresem $v_2^2$. Ale $\gcd(|v_1|, |v_2|) < |v_2|$ stąd sprzeczność z pierwotnością $v_2$.
+\end{proof}
+
+Dla słowa $x$, przez $p(x)$ oznaczamy najkrótszy okres $x$.
+
+\begin{lemma}\label{Lem_2}
+Niech $v_1, ..., v_r$ będzie sekwencją wszystkich $k$-HRP słowa $w$. Niech $[L_i, R_i]$ to zakres $v_i$. Wtedy dla każdego niepustego prefiksu $u$ słowa $w$ zachodzi:
+$$\begin{array}{rl}
+    p(u) = L_i/2 & \text{jeśli $|u| \in [L_i, R_i]$ dla pewnego $i$} \\
+    p(u) > |u|/k & \text{w przeciwnym przypadku}
+\end{array}$$
+\end{lemma}
+\begin{proof}
+Jeśli $|u|$ nie należy do żadnego przedziału $[L_i, R_i]$ to każdy okres $u$ musi być większy od $|u|/k$. 
+Jeśli $|u| \in [L_i, R_i]$ to $u$ musi być prefiksem $v_i^e$, dla $e \geq 2$. Z pierwotności $v_i$ najkrótszy okres $u$ to $|v_i| = L_i / 2$.
+\end{proof}
+
+Załóżmy, że $w$ jest wzorcem, dla którego $v_1$ jest jedynym $k$-HRP z zakresem $[L_1, R_1]$. Wtedy możemy znaleźć wszystkie wystąpienia $w$ w tekście $t$ w czasie liniowym.
+Niech $m = |w|, n = |t|$.
+
+\begin{algorithm}[H]
+\caption{SIMPLE-TEXT-SEARCH}
+\label{simple-text-search}
+\begin{algorithmic} 
+\STATE $pos \gets 0, j \gets 0$
+\WHILE{$pos + m \leq n$}
+    \WHILE{$j < m \land w[j + 1] = t[pos + j + 1] $}
+        \STATE $j \gets j + 1$
+    \ENDWHILE
+    \IF{$j = m$}
+        \RETURN Match at position $pos$.
+    \ENDIF
+    \IF{$j \in [L_1, R_1]$}
+        \STATE $pos \gets pos + L_1 / 2$
+        \STATE $j \gets j - L_1 / 2$
+    \ELSE
+        \STATE $pos \gets pos + \lfloor j / k \rfloor + 1$
+        \STATE $j \gets 0$
+    \ENDIF
+\ENDWHILE
+\end{algorithmic}
+\end{algorithm}
+Poprawność algorytmu wynika z lematu \ref{Lem_2}. Algorytm działa w czasie liniowym ponieważ cały czas rośnie wartość $k \cdot pos + j$. Łatwo też zauważyć, że dodatkowa pamięć wymagana przez algorytm jest stała.
+
+\begin{definition}
+\textbf{\textit{k-perfekcyjna-dekompozycja.}} Dla danego $w$ mówimy, że słowa $u,v$ są $k$-perfekcyjną dekompozycją jeśli $w = uv$, $v$ ma co najwyżej jedno $k$-HRP oraz $|u| < 2p(v)$.
+\end{definition}
+
+Zauważmy, że jeśli dla wzorca $w$ mamy $k$-perfekcyjną dekompozycję to możemy łatwo znaleźć wszystkie wystąpienia wzorca w czasie liniowym i stałej pamięci za pomocą algorytmu \ref{simple-text-search}. W czasie liniowym możemy wyszukać wszystkie wystąpienia $v$ i dla każdego sprawdzić czy jest poprzedzane przez $u$.
+Złożoność zamortyzuje się do czasu liniowego ponieważ $|u| \leq 2|v|$.
+Przedstawimy algorytm znajdujący $k$-perfekcyjną dekompozycję.
```

Comment:
"W czasie liniowym możemy wyszukać wszystkie wystąpienia $v$ i dla każdego sprawdzić czy jest poprzedzane przez $u$. Złożoność zamortyzuje się do czasu liniowego ponieważ $|u| \leq 2|v|$."
To be frank, I can't see it, can you elaborate on this?